### PR TITLE
Detecting duplicated array keys

### PR DESCRIPTION
**Pull request by rafalwrzeszcz**
This is the implementation fo #322 request.

I added it to `Design` section, any better idea?

### CODE REVIEW

**Comment by ravage84** [comment]
@rafalwrzeszcz thanks for the PR.

About the Ruleset, I think CleanCode or UnusedCode would be a better fit, as duplicated array keys have nothing to do with software design.

https://phpmd.org/rules/index.html

https://phpmd.org/rules/index.html#clean-code-rules
https://phpmd.org/rules/index.html#unused-code-rules

**Comment by ravage84** [comment]
AppVeyor fails:
https://ci.appveyor.com/project/phpmd/phpmd/build/57

````
There were 10 errors:
 
1) PHPMD\Rule\Design\DuplicatedArrayKeyTest::testRuleNotAppliesToMethodWithAssotiativeArrayDefinitionWithoutDuplicatedKeys
TypeError: Argument 1 passed to PHPUnit_Framework_TestCase::onNotSuccessfulTest() must be an instance of Exception, instance of Error given, called in C:\projects\phpmd\vendor\phpunit\phpunit\src\Framework\TestCase.php on line 851
 
2) PHPMD\Rule\Design\DuplicatedArrayKeyTest::testRuleAppliesToMethodWithAssotiativeArrayDefinitionWithDuplicatedKeys
TypeError: Argument 1 passed to PHPUnit_Framework_TestCase::onNotSuccessfulTest() must be an instance of Exception, instance of Error given, called in C:\projects\phpmd\vendor\phpunit\phpunit\src\Framework\TestCase.php on line 851
 
3) PHPMD\Rule\Design\DuplicatedArrayKeyTest::testRuleAppliesToMethodWithAssotiativeArrayDefinitionWithDuplicatedMixedTypeKeys
TypeError: Argument 1 passed to PHPUnit_Framework_TestCase::onNotSuccessfulTest() must be an instance of Exception, instance of Error given, called in C:\projects\phpmd\vendor\phpunit\phpunit\src\Framework\TestCase.php on line 851
 
4) PHPMD\Rule\Design\DuplicatedArrayKeyTest::testRuleAppliesToMethodWithAssotiativeArrayDefinitionWithDuplicatedMixedQuotedKeys
TypeError: Argument 1 passed to PHPUnit_Framework_TestCase::onNotSuccessfulTest() must be an instance of Exception, instance of Error given, called in C:\projects\phpmd\vendor\phpunit\phpunit\src\Framework\TestCase.php on line 851
 
5) PHPMD\Rule\Design\DuplicatedArrayKeyTest::testRuleAppliesMultipleTimesToMethodWithAssotiativeArrayDefinitionWithDuplicatedKeys
TypeError: Argument 1 passed to PHPUnit_Framework_TestCase::onNotSuccessfulTest() must be an instance of Exception, instance of Error given, called in C:\projects\phpmd\vendor\phpunit\phpunit\src\Framework\TestCase.php on line 851
 
6) PHPMD\Rule\Design\DuplicatedArrayKeyTest::testRuleNotAppliesToFunctionWithAssotiativeArrayDefinitionWithoutDuplicatedKeys
TypeError: Argument 1 passed to PHPUnit_Framework_TestCase::onNotSuccessfulTest() must be an instance of Exception, instance of Error given, called in C:\projects\phpmd\vendor\phpunit\phpunit\src\Framework\TestCase.php on line 851
 
7) PHPMD\Rule\Design\DuplicatedArrayKeyTest::testRuleAppliesToFunctionWithAssotiativeArrayDefinitionWithDuplicatedKeys
TypeError: Argument 1 passed to PHPUnit_Framework_TestCase::onNotSuccessfulTest() must be an instance of Exception, instance of Error given, called in C:\projects\phpmd\vendor\phpunit\phpunit\src\Framework\TestCase.php on line 851
 
8) PHPMD\Rule\Design\DuplicatedArrayKeyTest::testRuleAppliesToFunctionWithAssotiativeArrayDefinitionWithDuplicatedMixedTypeKeys
TypeError: Argument 1 passed to PHPUnit_Framework_TestCase::onNotSuccessfulTest() must be an instance of Exception, instance of Error given, called in C:\projects\phpmd\vendor\phpunit\phpunit\src\Framework\TestCase.php on line 851
 
9) PHPMD\Rule\Design\DuplicatedArrayKeyTest::testRuleAppliesToFunctionWithAssotiativeArrayDefinitionWithDuplicatedMixedQuotedKeys
TypeError: Argument 1 passed to PHPUnit_Framework_TestCase::onNotSuccessfulTest() must be an instance of Exception, instance of Error given, called in C:\projects\phpmd\vendor\phpunit\phpunit\src\Framework\TestCase.php on line 851
 
10) PHPMD\Rule\Design\DuplicatedArrayKeyTest::testRuleAppliesMultipleTimesToFunctionWithAssotiativeArrayDefinitionWithDuplicatedKeys
TypeError: Argument 1 passed to PHPUnit_Framework_TestCase::onNotSuccessfulTest() must be an instance of Exception, instance of Error given, called in C:\projects\phpmd\vendor\phpunit\phpunit\src\Framework\TestCase.php on line 851
 
FAILURES! 
Tests: 434, Assertions: 513, Errors: 10. 
````

**Comment by rafalwrzeszcz** [comment]
@ravage84 Moved to clean code. Any clue why AppVeyor is failing? The same test suite runs fine on Scrutinizer and Travis (and also locally).

It may be PHP7-related, but Travis build for PHP7 succeeds.

**Comment by ravage84** [comment]
Nope, no idea. Can you try to execute the tests on a Windows machine with PHP7?

File: src/test/resources/files/Rule/Design/DuplicatedArrayKey/testRuleNotAppliesToMethodWithoutArrayDefinition.php [hunk 1]
```diff
@@ -0,0 +1,8 @@
+<?php
+
+class testRuleNotAppliesToMethodWithoutArrayDefinition
```
Comment (ravage84) [inline src/test/resources/files/Rule/Design/DuplicatedArrayKey/testRuleNotAppliesToMethodWithoutArrayDefinition.php]: Not sure, if that's really necessary.

File: src/test/php/PHPMD/Rule/Design/DuplicatedArrayKeyTest.php [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+ * @group unittest
+ */
+class DuplicatedArrayKeyTest extends AbstractTest
+{
+    /**
+     * testRuleNotAppliesToMethodWithoutArrayDefinition
+     *
+     * @return void
+     */
+    public function testRuleNotAppliesToMethodWithoutArrayDefinition()
```
Comment (ravage84) [inline src/test/php/PHPMD/Rule/Design/DuplicatedArrayKeyTest.php]: Could you rename all the tests from ``xxxArrayDefinition`` to ``xxxArray``. Makes it a bit less long, still readable.

File: src/test/php/PHPMD/Rule/Design/DuplicatedArrayKeyTest.php [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+    {
+        $rule = new DuplicatedArrayKey();
+        $rule->setReport($this->getReportMock(0));
+        $rule->apply($this->getMethod());
+    }
+
+    /**
+     * testRuleNotAppliesToMethodWithNonAssotiativeArrayDefinition
```
Comment (ravage84) [inline src/test/php/PHPMD/Rule/Design/DuplicatedArrayKeyTest.php]: I know other tests aren't better, but could you make these test doc blocks more readable? Just having a 1:1 copy of the method name does not provide much additional info.

Either ``Test rule does not apply to method with non associative array``
or 
``Rule should not apply to method with non associative array``

Comment (ravage84) [inline src/test/php/PHPMD/Rule/Design/DuplicatedArrayKeyTest.php]: Tyo in ``assotiative``, should be ``associative``. Found 54 places with this typo.

**Comment by ravage84** [comment]
Would your code be able to detect duplicated array keys, if one would use a constant or class constant as the array key(s)?

For example:
````
define('FOO', 'foobar');
return array(
   FOO => 42,
   FOO => 43,
);
````
````
const FOO = 'foobar';
public function testBla() {
    return array(
       self::FOO => 42,
       self::FOO => 43,
    );
}
````
or mixed:
````
define('FOO', 'foobar');
return array(
   FOO => 42,
   'foobar' => 43,
);
````
````
const FOO = 'foobar';
public function testBla() {
    return array(
       self::FOO => 42,
       'foobar' => 43,
    );
}
````

**Comment by rafalwrzeszcz** [comment]
@ravage84 No and I don't think it's feasible in a reasonable time/PR size - constant can come from outside of the file or even a library.

But it should detect keys defined with same constant twice (like `self::FOO => 'one', self::FOO => 'two'`) as it checks the literal definition.

**Comment by ravage84** [comment]
That's fine for me. Just wanted to check.

**Comment by rafalwrzeszcz** [comment]
@ravage84 Then it has to wait few days, not sure if I will find some time to set up env for Win before the weekend.

**Comment by ravage84** [comment]
Take your time. I would like @manuelpichler to have an eye on this anyway. But I know he is **very** busy lately.

**Comment by ravage84** [comment]
@rafalwrzeszcz how is your status on this?

**Comment by ravage84** [comment]
@rafalwrzeszcz can you give me an update on this?

File: src/main/php/PHPMD/Rule/CleanCode/DuplicatedArrayKey.php [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+    private function analyzeArray(AbstractNode $node)
+    {
+        // small note regarding implementation - no need for recursion, as `apply()`
+        // finds all Array nodes on all depts level
+
+        $keys = array();
+        foreach ($node->findChildrenOfType('ArrayElement') as $arrayElement) {
+            // member of nested array - will be handled when `apply()` moves to that array
+            // could be nice if PDepend provides a method to fetch just direct children
+            if ($arrayElement->getParent() != $node) {
```
Comment (eeree) [inline src/main/php/PHPMD/Rule/CleanCode/DuplicatedArrayKey.php]: This one will end up with:
`PHP Fatal error:  Nesting level too deep - recursive dependency? in  src\main\php\PHPMD\Rule\CleanCode\DuplicatedArrayKey.php on line 87`
Comparing objects in PHP is recursive (deep-comparison).

File: src/main/php/PHPMD/Rule/CleanCode/DuplicatedArrayKey.php [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+                continue;
+            }
+
+            $children = $arrayElement->getChildren();
+            // non-associative array
+            if (count($children) == 1) {
```
Comment (eeree) [inline src/main/php/PHPMD/Rule/CleanCode/DuplicatedArrayKey.php]: This one could be improved, as given array:
```
$arr = [
    'foo',
    'bar',
    '' => 'baz',
    0 => '20',
    '0' => '30',
    false => '40',
    1 => '50',
    '1' => '60',
    true => '70',
    null => '80',
];
```

Will result in:

```
Array
(
    [0] => 40
    [1] => 70
    [] => 80
)
```
All due to casting.

File: src/main/php/PHPMD/Rule/CleanCode/DuplicatedArrayKey.php [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+     * @return Normalized key.
+     */
+    private function normalizeKey($key)
+    {
+        // in PHP keys are considered equal based on value
+        // so integer and string with same content will still point to same entry
+        // numbers don't need any processing, they have plain value
+
+        // this is string literal, not number
+        if (in_array($key[0], array('"', '\''))) {
```
Comment (eeree) [inline src/main/php/PHPMD/Rule/CleanCode/DuplicatedArrayKey.php]: It can be simplified to `trim($key, '\'"');`

**Comment by ravage84** [comment]
Closing as superseded by #484

**Pull request closed without merging**